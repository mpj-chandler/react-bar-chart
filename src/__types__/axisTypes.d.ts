import { SeriesData } from './seriesTypes';
import Padding from './stylingTypes';
import Placement from '../enums/Placement';

export interface AxisRange {
    min: number;
    max: number;
}

export interface AxisProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
    config?: AxisConfig;
}

export interface AxisConfig {
    zeroIntercept: boolean;
    margin: number;
    tickPlacement: Placement;
    tickLength: number;
}

export interface AxisTickProps {
    data: SeriesData[];
    padding: Padding;
    config?: AxisConfig;
}

