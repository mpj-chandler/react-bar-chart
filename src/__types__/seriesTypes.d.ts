import DataType from '../enums/DataType';
import { Padding } from './styling';
import { AxisRange } from './axisTypes';
import { Placement } from '../enums/Placement';

export interface DataPoint {
    tag?: string;
    y0?: number | null;
    y1: number | null;
}

export interface NamedDataPoint extends DataPoint {
    x: string;    
}

export interface DateIndexedDataPoint extends DataPoint {
    x: Date;
}

export interface NumericDataPoint extends DataPoint {
    x: number;
}

export interface NonNullNumericDataPoint extends NumericDataPoint {
    x: number;
}

export interface SeriesData {
    seriesName: string;
    type: {
        x: DataType,
        y: DataType,
    };
    points: Array<NamedDataPoint | DateIndexedDataPoint | NumericDataPoint | NonNullNumericDataPoint>;
}

export interface SeriesConfig {
    numSeries: number,
    seriesIndex: number,
    padding: Padding,
    dataType: DataType,
    xRange: AxisRange,
    yRange: any,
    placement: Placement
}