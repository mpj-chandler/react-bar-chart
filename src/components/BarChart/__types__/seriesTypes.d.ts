import DataType from "../enums/DataType";
import { Padding } from "./stylingTypes";
import { AxisRange } from "./axisTypes";
import Placement from "../enums/Placement";

export interface DataPoint {
    tag?: string;
}

export interface NamedDataPoint extends DataPoint {
    x: string;
    y: number;
}

export interface DateIndexedDataPoint extends DataPoint {
    x: Date;
    y: number;
}

export interface NumericDataPoint extends DataPoint {
    x: number | null;
    y: number | null;
}

export interface NonNullNumericDataPoint extends NumericDataPoint {
    x: number;
    y: number;
}

export interface SeriesData {
    seriesName: string;
    type: {
        x: DataType,
        y: DataType,
    };
    points: Array<NamedDataPoint | DateIndexedDataPoint | NumericDataPoint | NonNullNumericDataPoint>;
}

export interface SeriesConfig {
    numSeries: number,
    seriesIndex: number,
    padding: Padding,
    dataType: DataType,
    xRange: AxisRange,
    yRange: any,
    placement: Placement
}