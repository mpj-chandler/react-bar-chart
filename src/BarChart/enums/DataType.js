"use strict";
exports.__esModule = true;
var DataType;
(function (DataType) {
    DataType["Named"] = "NAMED";
    DataType["DateIndexed"] = "DATE_INDEXED";
    DataType["Numeric"] = "NUMERIC";
    DataType["NonNullNumeric"] = "NON_NULL_NUMERIC";
})(DataType || (DataType = {}));
;
exports["default"] = DataType;
