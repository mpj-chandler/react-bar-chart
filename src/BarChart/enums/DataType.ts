enum DataType {
    Named = 'NAMED',
    DateIndexed = 'DATE_INDEXED',
    Numeric = 'NUMERIC',
    NonNullNumeric = 'NON_NULL_NUMERIC',
};

export default DataType;
