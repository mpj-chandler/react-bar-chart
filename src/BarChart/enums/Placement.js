"use strict";
exports.__esModule = true;
var Placement;
(function (Placement) {
    Placement[Placement["Aligned"] = 0] = "Aligned";
    Placement[Placement["Bucket"] = 1] = "Bucket";
})(Placement || (Placement = {}));
exports["default"] = Placement;
