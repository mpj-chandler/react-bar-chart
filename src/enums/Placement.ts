enum Placement {
    Aligned,
    Bucket,
}

export default Placement;
