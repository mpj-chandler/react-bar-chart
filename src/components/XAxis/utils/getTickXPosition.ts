import { getXAxisRange } from '../../../utils/axisUtils/getXAxisRange/getXAxisRange';
import { Placement } from '../../../enums/Placement';
import { AxisTickProps } from '../../../__types__/axisTypes';
import DataType from '../../../enums/DataType';

export function getTickXPosition(
    props: AxisTickProps,
    dataType: DataType,
    labelConfig: { index: number, label: string }): number {

        switch (dataType) {
        case DataType.NonNullNumeric:
        case DataType.Numeric:
            const value = Number(labelConfig.label);

            if (isNaN(value)) {
                throw new Error('Error in calculating XAxis tick position! Label must be convertable to a number for label of this type!');
            }

            return getNumericTickXPosition(props, value);

        case DataType.Named:
            return getNumericTickXPosition(props, labelConfig.index);

        default:
            throw new Error('Error in calculating XAxis tick position! Data type not recognised!');
    }
}

function getNumericTickXPosition(props: AxisTickProps, value: number) {
    const range = getXAxisRange(props.data, props.config);
    const numPoints = props.data[0].points.length;
    const interval = range.max - range.min;
    const totalHorizontalPadding = props.padding.right + props.padding.left;

    if (props.config && props.config.tickPlacement === Placement.Bucket) {
        const step = interval / numPoints;

        return ((100 - totalHorizontalPadding) * ((step + value - range.min) / interval)) + props.padding.left;
    }

    return ((100 - totalHorizontalPadding) * (value - range.min) / interval) + props.padding.left;
}
