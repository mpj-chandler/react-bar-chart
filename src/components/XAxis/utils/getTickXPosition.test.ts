import { getTickXPosition } from './getTickXPosition';
import { AxisTickProps } from '../../../__types__/axisTypes';
import DataType from '../../../enums/DataType';
import Placement from '../../../enums/Placement';

describe('getTickXPosition', () => {
    describe('When using numeric x-value labels and default alignment', () => {
        const props: AxisTickProps = {
            data: [
                {
                    seriesName: 'A',
                    type: {
                        x: DataType.NonNullNumeric,
                        y: DataType.NonNullNumeric,
                    },
                    points: [
                        {
                            x: 0,
                            y: 10,
                        },
                        {
                            x: 1,
                            y: 20,
                        },
                        {
                            x: 2,
                            y: 30,
                        }
                    ],
                },
            ],
            padding: {
                top: 10,
                bottom: 10,
                left: 10,
                right: 10,
            },
        };

        it('it renders consistently', () => {
            const xPos = getTickXPosition(props, DataType.Numeric, { label: '8', index: 7 });

            expect(xPos).toMatchSnapshot();
        });

        it('it throws an error when a non-numeric value is passed', () => {
            expect(() => getTickXPosition(props, DataType.Numeric, { label: 'anything', index: 7 }))
                .toThrowErrorMatchingSnapshot();
        });
    });

    describe('When using numeric x-value labels and bucket alignment', () => {
        const props: AxisTickProps = {
            data: [
                {
                    seriesName: 'A',
                    type: {
                        x: DataType.NonNullNumeric,
                        y: DataType.NonNullNumeric,
                    },
                    points: [
                        {
                            x: 0,
                            y: 10,
                        },
                        {
                            x: 1,
                            y: 20,
                        },
                        {
                            x: 2,
                            y: 30,
                        },
                    ],
                },
            ],
            padding: {
                top: 10,
                bottom: 10,
                left: 10,
                right: 10,
            },
            config: {
                tickPlacement: Placement.Bucket,
                zeroIntercept: true,
                margin: 20,
                tickLength: 10,
            },
        };

        it('it renders consistently', () => {
            const xPos = getTickXPosition(props, DataType.Numeric, { label: '8', index: 7 });

            expect(xPos).toMatchSnapshot();
        });

        it('it throws an error when a non-numeric value is passed', () => {
            expect(() => getTickXPosition(props, DataType.Numeric, { label: 'anything', index: 7 }))
                .toThrowErrorMatchingSnapshot();
        });
    });

    describe('When using non-null numeric x-value labels and bucket alignment', () => {
        const props: AxisTickProps = {
            data: [
                {
                    seriesName: 'A',
                    type: {
                        x: DataType.NonNullNumeric,
                        y: DataType.NonNullNumeric,
                    },
                    points: [
                        {
                            x: 0,
                            y: 10,
                        },
                        {
                            x: 1,
                            y: 20,
                        },
                        {
                            x: 2,
                            y: 30,
                        },
                    ],
                },
            ],
            padding: {
                top: 10,
                bottom: 10,
                left: 10,
                right: 10,
            },
            config: {
                tickPlacement: Placement.Bucket,
                zeroIntercept: true,
                margin: 20,
                tickLength: 10,
            },
        };

        it('it renders consistently', () => {
            const xPos = getTickXPosition(props, DataType.NonNullNumeric, { label: '8', index: 7 });

            expect(xPos).toMatchSnapshot();
        });

        it('it throws an error when a non-numeric value is passed', () => {
            expect(() => getTickXPosition(props, DataType.Numeric, { label: 'anything', index: 7 }))
                .toThrowErrorMatchingSnapshot();
        });
    });

    describe('When using named x-value labels', () => {
        const props: AxisTickProps = {
            data: [
                {
                    seriesName: 'A',
                    type: {
                        x: DataType.Named,
                        y: DataType.NonNullNumeric,
                    },
                    points: [
                        {
                            x: 'A',
                            y: 10,
                        },
                        {
                            x: 'B',
                            y: 20,
                        },
                        {
                            x: 'C',
                            y: 30,
                        },
                    ],
                },
            ],
            padding: {
                top: 10,
                bottom: 10,
                left: 10,
                right: 10,
            },
        };

        it('it renders consistently', () => {
            const xPos = getTickXPosition(props, DataType.Named, { label: 'anything', index: 7 });

            expect(xPos).toMatchSnapshot();
        });
    });

    describe('When using an unknown label data type', () => {
        const props: AxisTickProps = {
            data: [
                {
                    seriesName: 'A',
                    type: {
                        x: DataType.DateIndexed,
                        y: DataType.NonNullNumeric,
                    },
                    points: [
                        {
                            x: 'A',
                            y: 10,
                        },
                        {
                            x: 'B',
                            y: 20,
                        },
                        {
                            x: 'C',
                            y: 30,
                        }
                    ],
                },
            ],
            padding: {
                top: 10,
                bottom: 10,
                left: 10,
                right: 10,
            },
        };

        it('it throws an error', () => {
            expect(() => getTickXPosition(props, DataType.DateIndexed, { label: 'anything', index: 7 }))
                .toThrowErrorMatchingSnapshot();
        });
    });
});

