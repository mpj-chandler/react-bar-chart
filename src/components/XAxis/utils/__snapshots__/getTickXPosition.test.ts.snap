// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTickXPosition When using an unknown label data type it throws an error 1`] = `"Error in calculating XAxis tick position! Data type not recognised!"`;

exports[`getTickXPosition When using named x-value labels it renders consistently 1`] = `290`;

exports[`getTickXPosition When using non-null numeric x-value labels and bucket alignment it renders consistently 1`] = `236.66666666666669`;

exports[`getTickXPosition When using non-null numeric x-value labels and bucket alignment it throws an error when a non-numeric value is passed 1`] = `"Error in calculating XAxis tick position! Label must be convertable to a number for label of this type!"`;

exports[`getTickXPosition When using numeric x-value labels and bucket alignment it renders consistently 1`] = `236.66666666666669`;

exports[`getTickXPosition When using numeric x-value labels and bucket alignment it throws an error when a non-numeric value is passed 1`] = `"Error in calculating XAxis tick position! Label must be convertable to a number for label of this type!"`;

exports[`getTickXPosition When using numeric x-value labels and default alignment it renders consistently 1`] = `330`;

exports[`getTickXPosition When using numeric x-value labels and default alignment it throws an error when a non-numeric value is passed 1`] = `"Error in calculating XAxis tick position! Label must be convertable to a number for label of this type!"`;
